app/src/main/java/org/mozilla/rocket/content/news/NewsViewModel.kt

```diff
@@ -1,27 +1,62 @@
 package org.mozilla.rocket.content.news
 
-import android.arch.lifecycle.MutableLiveData
+import android.arch.lifecycle.LiveData
+import android.arch.lifecycle.MediatorLiveData
+import android.arch.lifecycle.Transformations
 import android.arch.lifecycle.ViewModel
 import org.mozilla.lite.partner.NewsItem
 import org.mozilla.lite.partner.Repository
+import org.mozilla.rocket.content.Result
 
-class NewsViewModel : ViewModel(), Repository.OnDataChangedListener<NewsItem> {
-    var repository: Repository<out NewsItem>? = null
-        set(value) {
-            if (field != value) {
-                items.value = null
+class NewsViewModel(
+    private val loadNewsCategoryUseCase: LoadNewsCategoryUseCase
+) : ViewModel() {
+
+    private var newCategoryResult: LiveData<Result<LoadNewsCategoryByLangResult>> =
+        loadNewsCategoryUseCase.observe()
+
+    val categories: LiveData<List<String>> =
+        Transformations.map(this.newCategoryResult) { (it as? Result.Success)?.data?.categories }
+
+    private val newsMap = HashMap<String, MediatorLiveData<List<NewsItem>>>()
+
+    private val useCaseMap = HashMap<String, LoadNewsUseCase>()
+
+    private val queryMap = HashMap<String, HashMap<String, String>>()
+
+    init {
+        updateCategory("english")
+    }
+
+    private fun updateCategory(input: String) {
+        val param = LoadNewsCategoryByLangParameter(input)
+        loadNewsCategoryUseCase.execute(param)
+    }
+
+    fun getNews(category: String, lang: String, repo: Repository<out NewsItem>): MediatorLiveData<List<NewsItem>> {
+        val items = newsMap[category] ?: MediatorLiveData()
+        if (newsMap[category] == null) {
+            newsMap[category] = items
+            queryMap[category] = HashMap<String, String>().apply {
+                this["category"] = category
+                this["lang"] = lang
+            }
+            val loadNewsCase = LoadNewsUseCase(repo)
+            useCaseMap[category] = loadNewsCase
+            items.addSource(loadNewsCase.observe()) { result ->
+                (result as? Result.Success)?.data?.let {
+                    items.value = it.items
+                }
             }
-            field = value
         }
-    val items = MutableLiveData<List<NewsItem>>()
 
-    override fun onDataChanged(newsItemList: List<NewsItem>?) {
-        // return the new list, so diff utils will think this is something to diff
-        items.value = newsItemList
+        return items
     }
 
-    fun loadMore() {
-        repository?.loadMore()
-        // now wait for OnDataChangedListener.onDataChanged to return the result
+    fun loadMore(category: String) {
+        useCaseMap[category]?.execute(parse(queryMap[category]!!))
```

Comment:
Do we really need to have `queryMap`? just keep single language variable should be enough?